eo-runtime/src/main/java/EOorg/EOeolang/EOcage.java

```diff
@@ -108,4 +112,96 @@ public Phi lambda() {
             return new Data.ToPhi(true);
         }
     }
+
+    /**
+     * Class to trace if the cage got into recursion during the dataization.
+     * @since 0.36
+     * @todo #2836:60min Add a new parameter of recursion depth. This parameter
+     *   should be set by user via pom.xml. We can make DATAIZING_CAGES a
+     *   ConsurrentHashMap which will add thread safety to the class.
+     */
+    private static final class PhTracedEnclosure implements Phi {
+
+        /**
+         * Cages that are currently dataizing. If one cage is datazing and
+         * it needs to be dataized inside current dataizing, the cage will be here,
+         */
+        private static final Set<Integer> DATAIZING_CAGES = new HashSet<>();
+
+        /**
+         * Enclosure.
+         */
+        private final Phi enclosure;
+
+        /**
+         * Vertex of cage where the {@link PhTracedEnclosure#enclosure}
+         * was retrieved.
+         */
+        private final int cage;
+
+        /**
+         * Ctor.
+         * @param enclosure Enclosure.
+         * @param cage Vertex of source cage.
+         */
+        PhTracedEnclosure(final Phi enclosure, final int cage) {
+            this.enclosure = enclosure;
+            this.cage = cage;
+        }
+
+        @Override
+        public Phi copy() {
+            return new PhTracedEnclosure(this.enclosure, this.cage);
+        }
+
+        @Override
+        public String locator() {
+            return this.enclosure.locator();
+        }
+
+        @Override
+        public String forma() {
+            return this.enclosure.forma();
+        }
+
+        @Override
+        public String φTerm() {
+            return this.enclosure.φTerm();
+        }
+
+        @Override
+        public Attr attr(final int pos) {
+            return this.getAttrSafely(() -> this.enclosure.attr(pos));
+        }
+
+        @Override
+        public Attr attr(final String name) {
+            return this.getAttrSafely(() -> this.enclosure.attr(name));
+        }
+
+        @Override
+        public int hashCode() {
+            return this.enclosure.hashCode();
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            return obj instanceof Phi && this.hashCode() == obj.hashCode();
```

Comment:
@levBagryansky [Optional]  it might be dangerous part of the code. Some objects might not be equal, but have the same `hashCode()`:
```java
System.out.println(Objects.hashCode(126L) == Objects.hashCode(4294967423L));
//prints 'true'
```
Maybe we can leave `todo` here to fix it later, what do you think?

Reply:
@volodya-lombrozo as for `Phi` we have an important condition that `Phi` with different `hashcode` are different. This is written in `Phi` documentation: https://github.com/objectionary/eo/blob/master/eo-runtime/src/main/java/org/eolang/Phi.java#L34

Reply:
@levBagryansky Well, it's a bit strange, but ok. Thank you for the explanation.